### PR TITLE
Vertex density & quad overdraw debug modes

### PR DESCRIPTION
**Pull request by adrien-de-tocqueville**
### Purpose of this PR
This PR is adding two fullscreen debug modes in rendering pannel (not available on mac):
- Quad Overdraw: Displays an overlay that highlights pixels running multiple fragment shaders.
quad

- Vertex density:  Displays an overlay that highlights pixels containing multiple vertices. Helps finding models that need LODs. 
density

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [x] Checked new UI names with UX convention
- [x] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [x] C# and shader warnings (supress shader cache to see them)
- [x] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

**Yamato**: (Select your branch):
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Fvertex-density

Adds two graphics tests (3004 for quad overdraw & 3005 for vertex density) which have all standard and SG materials variations

### CODE REVIEW

**Comment by adrien-de-tocqueville** [comment]
> With Max Vertex Density <5 render mode gives inconsistent results. You can see that some vertices are assigned random colors while having a homogeneous wireframe. According to Adrien these are shader shenanigans, where GPU can recalculate random vertices.

To be precise, that's not random, that depends on the vertex order of the model. It may show that reordering the vertices can lead to fewer vertex shader execution

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl [hunk 1]
```diff
@@ -178,6 +184,36 @@ void Frag(PackedVaryingsToPS packedInput,
             float4 result = _DebugTransparencyOverdrawWeight * float4(TRANSPARENCY_OVERDRAW_COST, TRANSPARENCY_OVERDRAW_COST, TRANSPARENCY_OVERDRAW_COST, TRANSPARENCY_OVERDRAW_A);
             outColor = result;
         }
+        else if (_DebugFullScreenMode == FULLSCREENDEBUGMODE_QUAD_OVERDRAW)
+        {
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl:188]: Please reference the original article: https://blog.selfshadow.com/2012/11/12/counting-quads/

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl [hunk 2]
```diff
@@ -60,6 +60,9 @@ PackedVaryingsToPS VertTesselation(VaryingsToDS input)
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/TessellationShare.hlsl"
 #endif
 
+#if defined(DEBUG_DISPLAY) && !defined(_DEPTHOFFSET_ON)
+[earlydepthstencil] // quad overshading debug mode write to UAVs
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl:64]: Do we know if attribute really work with guarded define - i.e arent they always apply even if there is guard ? 
(poking @JulienIgnace-Unity  / @FrancescoC-unity )

File: com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs [hunk 3]
```diff
@@ -113,6 +116,8 @@ public void InitSharedBuffers(GBufferManager gbufferManager, RenderPipelineSetti
                 // In case of deferred, we must be in sync with NormalBuffer.hlsl and lit.hlsl files and setup the correct buffers
                 m_NormalRT = gbufferManager.GetNormalBuffer(0); // Normal + Roughness
             }
+
+            m_DebugDisplayUAV = RTHandles.Alloc(Vector2.one, TextureXR.slices, colorFormat: GraphicsFormat.R32_UInt, dimension: TextureXR.dimension, enableRandomWrite: true, useDynamicScale: true, name: "DebugDisplayUAV");
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs:120]: @JulienIgnace-Unity we really need to do a pass on memory (note poking you also for render graph).
Currently we allocate a lot of memory even if we don't used thing, I know render graph can help here. But right now situation is still not good.
For the memory above for example 1) we shouldn't allocated in release or if we don't require debug, 2) we could reuse m_ContactShadowBuffer memory (which shouldn't be allocated if we disble shadow).

We will still merge this PR with this, but we really need to do a better memory management

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl (hunk 2 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl:64]: Gaenerally the preprocessor is run before the attributes are evaluated for compilation (made small example with [branch] here http://shader-playground.timjones.io/82ad51ba693770094363e31f8b3bafe6 you can see in the DXC output how the branch contains the hint or not depending on the define value), so I think this should work fine, though never tried myself with forcing early depth stencil.

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.hlsl [hunk 4]
```diff
@@ -17,6 +17,12 @@ static DirectionalShadowType g_DebugShadowAttenuation = 0;
 
 StructuredBuffer<int2>  _DebugDepthPyramidOffsets;
 
+// For quad overdraw, we need two halfscreen UAV (because a quad is 2x2 pixels). For vertex density, we need one fullscreen UAV.
+// These modes are exclusive so we make only one fullscreen allocation for both.
+// For vertex density, it stores the number of vertex projected in each pixel.
+// For quad overdraw, each 2x2 quad of the UAV contains the overdraw count in top-left pixel and the locked quad in the bottom-right pixel. The two other pixels of the quad are unused.
+RW_TEXTURE2D_X(uint, _DebugDisplayUAV) :register(u2);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.hlsl:24]: @sebastienlagarde  I don't remember the details of our platforms implementation, but for the explicit registers in the uav did it _need_ to be starting from after render target count or it is fine as soon as the register is _after_ the render target count?

File: com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs (hunk 3 above)
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs:120]: Agreed but the question is: Do we want to spend time now doing this for the current code path are should we "just" wait for Render Graph to be enabled by default? Since we target Render Graph for the end of the month I'd say we wait.

File: com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs [hunk 5]
```diff
@@ -160,6 +165,12 @@ public RenderTargetIdentifier[] GetMotionVectorsPassBuffersRTI(FrameSettings fra
             }
         }
 
+        // Function that will return the UAV for quad overshading and vertex density debug modes
+        public RenderTargetIdentifier GetDebugDisplayBufferRTI()
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs:169]: I'm not sure this is necessary. RTHandle should convert implicitly to RenderTargetIdentifier iirc.

Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/Material/SharedRTManager.cs:169]: I mean, we can just return the RTHandle and be consistent with most of our other APIs. In general the only place we need the RTI is for MRT arrays.

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.hlsl (hunk 4 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.hlsl:24]: yes, I have think about that as well, guess the code may not work on PS4.
on PS4 we will need to bing on 2 + number of RT allocated. We will check in a second time (I tracked this).

**Comment by sebastienlagarde** [comment]
Note: not sure yet but the PR could affect Vulkan (guess this could be related to the early stencil?). I have re-run vulkan to be sure.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl [hunk 6]
```diff
@@ -140,9 +140,9 @@ VaryingsMeshType VertMesh(AttributesMesh input)
     output.normalWS = normalWS;
     output.tangentWS = tangentWS;
 #endif
-#if defined(DEBUG_DISPLAY) && (SHADERPASS == SHADERPASS_FORWARD || SHADERPASS == SHADERPASS_FORWARD_UNLIT)
-        if (_DebugFullScreenMode == FULLSCREENDEBUGMODE_VERTEX_DENSITY)
-            IncrementVertexDensityCounter(output.positionCS);
+#if defined(SHADERPASS) && (SHADERPASS == SHADERPASS_FULLSCREEN_DEBUG)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl:143]: you don't need to check for #if defined(SHADERPASS) 
you can just check
#if SHADERPASS == SHADERPASS_FULLSCREEN_DEBUG

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader [hunk 7]
```diff
@@ -804,6 +804,37 @@ Shader "HDRP/LitTessellation"
 
             ENDHLSL
         }
+
+        Pass
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader:808]: still reamin LayerdLit, LayeredLitTessellation, Axf and shader graph, but I get first you validate with this

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 8]
```diff
@@ -1241,14 +1244,16 @@ void Resize(HDCamera hdCamera)
...
                     m_SharedRTManager.DisposeCoarseStencilBuffer();
                 }
 
                 LightLoopAllocResolutionDependentBuffers(hdCamera, m_MaxCameraWidth, m_MaxCameraHeight);
+                m_SharedRTManager.AllocateFullScreenDebugBuffer(m_MaxCameraWidth, m_MaxCameraHeight, m_MaxViewCount);
                 if (!m_EnableRenderGraph)
                 {
                     m_DbufferManager.AllocResolutionDependentBuffers(m_MaxCameraWidth, m_MaxCameraHeight);
-                    m_SharedRTManager.AllocateCoarseStencilBuffer(m_MaxCameraWidth, m_MaxCameraHeight, hdCamera.viewCount);
+                    m_SharedRTManager.AllocateCoarseStencilBuffer(m_MaxCameraWidth, m_MaxCameraHeight, m_MaxViewCount);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1256]: Why this change ?m_MaxViewCount

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs [hunk 9]
```diff
@@ -135,11 +135,13 @@ internal List<(Camera, XRPass)> SetupFrame(Camera[] cameras, bool singlePassAllo
                     // Disable vsync on the main display when rendering to a XR device
                     QualitySettings.vSyncCount = 0;
 
+#if ENABLE_VR && ENABLE_XR_MODULE
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs:138]: hmm... Why do you need this, it doesn't compule? (cc @fabien-unity )

Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs:138]: I already fixed this here https://github.com/Unity-Technologies/Graphics/pull/1391/files  (please revert this bit as it will be redundant and it won't merge well automatically)

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl (hunk 6 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl:143]: i didn't look too much into it but it looks like the file is included somewhere SHADERPASS is not defined

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs (hunk 8 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1256]: Because that could lead to an incorrect allocation in some cases where you change resolution and view count in a specific order (because no reallocation is done when hdCamera.viewCount is inferior to m_MaxViewCount)

**Comment by adrien-de-tocqueville** [comment]
Compilation error on metal
> error: use of undeclared identifier 'input'
> u_xlati6 = int(input.SV_PrimitiveID0) + 0x1;

Looks like a bug when the shader gets converted from HLSL to MSL.
SV_PrimitiveID is necessary for the quad overdraw debug mode only

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl [hunk 10]
```diff
@@ -177,7 +177,13 @@ VaryingsMeshToPS VertMeshTesselation(VaryingsMeshToDS input)
     output.positionCS = TransformWorldToHClip(input.positionRWS);
 
 #if defined(SHADERPASS) && (SHADERPASS == SHADERPASS_FULLSCREEN_DEBUG)
+#ifdef SHADER_API_METAL
+    // There is a bug with generated shader on metal when accessing the ShaderVariablesDebugDisplay cbuffer in this shader stage.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl:181]: I am not sure to understand the workaround, why not doing 

#ifndef SHADER_API_METAL 
#if defined(SHADERPASS) && (SHADERPASS == SHADERPASS_FULLSCREEN_DEBUG)
if (_DebugFullScreenMode == FULLSCREENDEBUGMODE_VERTEX_DENSITY)
 IncrementVertexDensityCounter(output.positionCS);
#endif
#"endif?

or simpler, inside IncrementVertexDensityCounter you can do the test on SHADER_API_METAL

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs (hunk 8 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1250]: @fabien-unity isn't this a problem for VR?
@adrien-de-tocqueville so it is a bug unrelated to this PR?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl (hunk 10 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/VertMesh.hlsl:181]: This bug is for vertex density, not quad overdraw, and it should work on mac. so with what I did it is supposed to work also on mac.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs (hunk 8 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1250]: Yes it's unrelated, i just saw it because I had to copy the line

Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1250]: > @fabien-unity isn't this a problem for VR?
all good. this will avoid issues when switching between VR and non-VR modes.

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-10-0.md [hunk 11]
```diff
@@ -112,6 +112,11 @@ For more information, see the Lighting panel section in the [HDRP debug window](
 
 The Render Pipeline Debug window now has a new Volume panel which you can use to visualize the Volume components that affect a specific Camera. For each Volume that contributes to the final interpolated value, this panel shows the value of each property and whether or not it is overridden. It also calculates the Volume's influence percentage using the Volume's weight and blend distance. For more information, see the Volume panel section in the [HDRP debug window](Render-Pipeline-Debug-Window.md#VolumePanel).
 
+#### Quad Overdraw and Vertex Density
+
+Quad Overdraw highlights GPU quads running multiple fragment shaders, which is mainly caused by small or thin triangles while Vertex Density displays pixels running multiple vertex shaders.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Documentation~/whats-new-10-0.md:117]: please add that it is not supported on metal

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs [hunk 12]
```diff
@@ -1585,7 +1632,7 @@ void RegisterRenderingDebug()
                 widgetList.Add(new DebugUI.BoolField { displayName = "XR single-pass test mode", getter = () => data.xrSinglePassTestMode, setter = value => data.xrSinglePassTestMode = value });
             }
 
-            // widgetList.Add(new DebugUI.BoolField { displayName = "Enable Render Graph", getter = () => HDRenderPipeline.currentPipeline.IsRenderGraphEnabled(), setter = value => HDRenderPipeline.currentPipeline.EnableRenderGraph(value) });
+             widgetList.Add(new DebugUI.BoolField { displayName = "Enable Render Graph", getter = () => HDRenderPipeline.currentPipeline.IsRenderGraphEnabled(), setter = value => HDRenderPipeline.currentPipeline.EnableRenderGraph(value) });
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:1635]: this shouldn't be part of the PR (julien will have a separate PR for this).

Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:1635]: oops, i forgot to remove it

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-10-0.md (hunk 11 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Documentation~/whats-new-10-0.md:117]: it's already there in the line under

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitDataMeshModification.hlsl [hunk 13]
```diff
@@ -114,7 +114,7 @@ float4 GetTessellationFactors(float3 p0, float3 p1, float3 p2, float3 n0, float3
         edgeTessFactors *= distFactor * distFactor;
     }
 
-    edgeTessFactors *= _TessellationFactor;
+    edgeTessFactors *= _TessellationFactor * _GlobalTessellationFactorMultiplier;
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitDataMeshModification.hlsl:117]: Does not seem to be relared to this PR?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 14]
```diff
@@ -4122,6 +4112,55 @@ void RenderTransparencyOverdraw(CullingResults cull, HDCamera hdCamera, Scriptab
...
+        {
+            var parameters = new FullScreenDebugParameters();
+
+            parameters.rendererList = CreateOpaqueRendererListDesc(cull, hdCamera.camera, m_FullScreenDebugPassNames, renderQueueRange: RenderQueueRange.all);
+            parameters.frameSettings = hdCamera.frameSettings;
+
+            return parameters;
+        }
+
+        void RenderFullScreenDebug( FullScreenDebugParameters   parameters,
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:4131]: This function should be static otherwise when called from a render graph pass, it will capture "this" and generate GCAlloc.
To check this, you can hover the mouse over the => of the lambda calling this and check that it does not capture anything.

**Review by JulienIgnace-Unity — CHANGES_REQUESTED** [review]
RenderFullScreenDebug needs to be static. Otherwise lgtm.
Is it expected that we have so many scenes, screenshots and prefab modified? Should this change any test at all?

**Comment by adrien-de-tocqueville** [comment]
> Is it expected that we have so many scenes, screenshots and prefab modified? Should this change any test at all?

There are two new tests but nothing else have been modified

**Comment by JulienIgnace-Unity** [comment]
My bad, did not see most of them are in the new scenes. There is just one file for the DebugLightLayer scene then that should probably not be changed (probably not a big deal)

**Comment by adrien-de-tocqueville** [comment]
> There is just one file for the DebugLightLayer

It's because I changed the script to set the debug mode for the tests to be more consistent

**Review by JulienIgnace-Unity — APPROVED** [review]

**Comment by PaulDemeulenaere** [comment]
Note, this change fails with PSSL as well : SV_PrimitiveID isn't available at fragment stage.
We are fixing this asap.
